core/services/workflows/engine.go

```diff
@@ -866,3 +883,133 @@ func executeSyncAndUnwrapSingleValue(ctx context.Context, cap capabilities.Callb
 
 	return l.Underlying[0], nil
 }
+
+// Logging keys
+const (
+	capabilityIDKey = "capabilityID"
+	triggerIDKey    = "triggerID"
+	workflowIDKey   = "workflowID"
+	executionIDKey  = "executionID"
+	stepRefKey      = "stepRef"
+)
+
+// We don't define Unwrap since we don't want to expose the underlying error.
+type WorkflowError struct {
```

Comment:
Discussed offline, consider something like the following to reduce the number of error types we're introducing here:

```
type workflowError struct {
    keys map[string]any
    innerErr error
}

func (w workflowError) Error() string {
    ...
}

newStepError(stepRef string, executionID string) workflowError

newExecutionError(keys) workflowError

```

if you want to make this wrappable, you can include `innerErr` and unwrap that.